src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.InternalAbstractions;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.Internal;
+using NuGet.Frameworks;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Internal
+{
+    public class PrecompileDispatchCommand : PrecompileCommandBase
+    {
+        private CommandOption NoBuildOption { get; set; }
+
+        protected override void Configure(CommandLineApplication app)
+        {
+            base.Configure(app);
+            NoBuildOption = app.Option(
+                "--no-build",
+                "Do not build project before compiling views.",
+                CommandOptionType.NoValue);
+        }
+
+        protected override int ExecuteCore()
+        {
+            if (!NoBuildOption.HasValue())
+            {
+                var exitCode = BuildProject();
+                if (exitCode != 0)
+                {
+                    return exitCode;
+                }
+            }
+
+            var dispatchArgs = new List<string>
+            {
+                ProjectPath,
+                "--framework",
+                TargetFramework.ToString(),
+                "--configuration",
+                Configuration,
+            };
+
+            string outputPath = null;
+            if (OutputPathOption.HasValue())
+            {
+                outputPath = OutputPathOption.Value();
+
+                dispatchArgs.Add("--output");
+                dispatchArgs.Add(outputPath);
+            }
+
+            if (ConfigureCompilationType.HasValue())
+            {
+                dispatchArgs.Add("--configure-compilation-type");
+                dispatchArgs.Add(ConfigureCompilationType.Value());
+            }
+
+            if (ContentRootOption.HasValue())
+            {
+                dispatchArgs.Add("--content-root");
+                dispatchArgs.Add(ContentRootOption.Value());
+            }
+
+            if (GeneratePdbOption.HasValue())
+            {
+                dispatchArgs.Add("--generate-pdbs");
+            }
+
+            var toolName = typeof(Precompilation.Program).GetTypeInfo().Assembly.GetName().Name;
+            var dispatchCommand = DotnetToolDispatcher.CreateDispatchCommand(
+                dispatchArgs,
+                TargetFramework,
+                Configuration,
+                outputPath: outputPath,
+                buildBasePath: null,
```

Comment:
The output path listed in xproj is actually used as the build-base-path. Super confusing I know. Even filed an issue but I guess it's by design? :/ https://github.com/aspnet/Tooling/issues/455